config/http_config.go

```diff
@@ -769,18 +883,28 @@ func (rt *oauth2RoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 		changed bool
 	)
 
-	if rt.clientSecret != nil {
-		var err error
-		secret, err = rt.clientSecret.fetch()
-		if err != nil {
-			return nil, fmt.Errorf("unable to read oauth2 client secret: %w", err)
+	// Fetch the secret if it's our first run or always if the secret can change.
+	if rt.rt == nil || (rt.clientSecret != nil && !rt.clientSecret.immutable()) {
```

Comment:
This not ended up better (or worse) after my suggestions -- might be my misunderstanding, so I ended up prototyping how it might look, proposed in https://github.com/TheSpiritXIII/prometheus-common/pull/1

I also fixed a small race with reading `rt.rt` 🤔I think this race was before too.

Reply:
No regression and I only refactor in my PR, so let's merge this one and review my PR if there is a need (https://github.com/prometheus/common/blob/58f58f894eb3dd8aafd868824747a997c6fb6e83/config/http_config.go#L759)